### PR TITLE
Add a new endpoint that exposes block editor settings through the REST API

### LINKED ISSUE #29063: Mobile - Add new endpoint to get global styles settings through the REST API
## What problem does this address?
In order to add support for Global Styles on the mobile editor, there should be an endpoint to be able to fetch [this data](https://github.com/WordPress/gutenberg/blob/master/lib/global-styles.php#L191) through the REST API.

## What is your proposed solution?
There's an [opened PR](https://github.com/WordPress/gutenberg/pull/25226) with an exploration of this so we'd need to either continue that effort to use it as a reference.

### PR DESCRIPTION
**Pull request by geriux**
Fixes https://github.com/WordPress/gutenberg/issues/29063

## Description
Adds a new REST API endpoint to expose block editor settings. Useful to get things like global style settings and more.

## How has this been tested?

Using a local instance of WordPress:

- Use postman to call `/__experimental/wp-block-editor/v1/settings` 
- TBD (In progress)

## Screenshots 
N/A

## Types of changes
New rest API endpoint.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by geriux** [comment]
Hey @nosolosw 👋 Since you are currently working on Global styles, I'd appreciate any feedback you might have regarding this PR. Thank you so much!

**Comment by gziolo** [comment]
It would probably be better to build this REST API endpoint in a way that allows using it on the front-end by Gutenberg or any other projects that needs that detail.

**Comment by geriux** [comment]
> It would probably be better to build this REST API endpoint in a way that allows using it on the front-end by Gutenberg or any other projects that needs that detail.

Hey @gziolo thank you for your feedback! I changed the naming a bit so it's just for getting global styles settings in general and removing any mobile-specific mention.

**Comment by TimothyBJacobs** [comment]
`wp/v2` is really for modeling core entities. I wonder if it might make sense to put pure block editor endpoints in something like `wp-block-editor/v1` matching what we did for Site Health.

File: lib/load.php
Comment (enejb) [inline lib/load.php]: Lets use `require_once` here so that the file can't be loaded multiple times for some reason.

Comment (geriux) [reply lib/load.php]: Thanks for the suggestion! I'll change it 👍

**Comment by gziolo** [comment]
One thing that I'm not sure about is why you think that the settings for Global Styles should be consumed separately from all other settings. In particular, I'm worried that `gutenberg_global_styles_rest_settings` duplicates what already exists for the editor. Ideally, we should have one shared set of setting that we can one day consume through the same endpoint on the web without too much work with mapping.

**Comment by oandregal** [comment]
:wave: I think I miss some context into what mobile wants to do to be more helpful. To offer some feedback, I'm assuming that this is what mobile wants:

- mobile will like to use the existing global configs (style & settings)
- mobile won't need to be able to set those *global* configs (this is what we do in the site editor via the global styles sidebar)

Assuming that's the goal, here's some feedback:

- If the user can't change the global config from mobile, the endpoint could expose the consolidated settings (core, theme, user) instead of exposing the user entity id + the base configs.

- Can we perhaps expose through the endpoint the block editor settings instead? Or is there anything special that mobile needs? The only data I see different from block editor settings is that this endpoint also exposes whether the theme has support for theme.json. Why is this needed in mobile? This issue may be relevant https://github.com/WordPress/gutenberg/issues/30082 (we're trying to see how to consolidate these settings).

- Another thought that I have is that the theme.json shape is going to need evolving. I'm working on a PR that will introduce that to be able to do these changes https://github.com/WordPress/gutenberg/issues/29891 Not sure how this affects the REST endpoint.

**Comment by geriux** [comment]
Thank you for the feedback @gziolo and @nosolosw 

> One thing that I'm not sure about is why you think that the settings for Global Styles should be consumed separately from all other settings. In particular, I'm worried that gutenberg_global_styles_rest_settings duplicates what already exists for the editor. Ideally, we should have one shared set of setting that we can one day consume through the same endpoint on the web without too much work with mapping.

> Can we perhaps expose through the endpoint the block editor settings instead? Or is there anything special that mobile needs?

Ideally it would be amazing to just expose the block editor settings, but I encountered some issues because within `gutenberg_experimental_global_styles_settings` there are things that are not available when applying the block editor settings filter from the endpoint, like  `get_current_screen`, `function_exists( 'gutenberg_is_edit_site_page' )` and `gutenberg_is_edit_site_page( $screen->id )` I guess we could pass some sort of flag to not check this? Do we have any cases like this one?

> 👋 I think I miss some context into what mobile wants to do to be more helpful. To offer some feedback, I'm assuming that this is what mobile wants

From mobile we'd need:

- Active theme colors and gradients (for cases where there is no theme.json). The current `__experimentalFeatures` does not return the active theme custom colors (`editor-color-palette` and `editor-gradient-presets`), just the default ones from the block editor.
- Global styles settings: we are starting with root styles (to customize the editor with the theme colors) and then continuing with block styles.

> If the user can't change the global config from mobile, the endpoint could expose the consolidated settings (core, theme, user) instead of exposing the user entity id + the base configs.

For that you mean using `WP_Theme_JSON_Resolver::get_merged_data` right? That works to have the colors data but we'd also need things like:

```json
{
    "styles": {
        "root": {
            "color": {
                "background": "var(--wp--preset--color--green)",
                "text": "var(--wp--preset--color--dark-gray)",
                "link": "var(--wp--preset--color--dark-gray)"
            }
        }
    }
}
````

To be able to customize the mobile editor's background, text, and links with the theme style.

I guess we can exclude the user entity id for now since we are not supporting the possibility to change the config just yet.

Thanks again for the help!

File: lib/class-wp-rest-block-editor-global-styles-settings-controller.php
Comment (spacedmonkey) [inline lib/class-wp-rest-block-editor-global-styles-settings-controller.php]:
```suggestion
		$this->namespace = 'wp/v2';
```

Comment (geriux) [reply lib/class-wp-rest-block-editor-global-styles-settings-controller.php]: Hey there 👋 the initial PR had this naming but it was [suggested](https://github.com/WordPress/gutenberg/pull/29969#issuecomment-802137951) to change it to follow a previously implemented endpoint for the block editor. What do you think? Thanks!

**Comment by geriux** [comment]
Hey there 👋 

I've updated the code to use the current `block_editor_settings` instead, what do you think of this approach?

Also, @nosolosw wanted to ask if it'd be possible for `__experimentalFeatures` to return the active theme custom colors from `editor-color-palette` and `editor-gradient-presets` (when there's any) instead of returning the default values from `get_core_data`. Similar to when you have for example `TT1 blocks` activated and it returns the correct color palette.

Thanks!!

File: lib/global-styles.php
Comment (geriux) [inline lib/global-styles.php]: These changes are pending to be updated after this https://github.com/WordPress/gutenberg/issues/30082#issuecomment-816679803

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (geriux) [inline lib/class-wp-rest-block-editor-settings-controller.php]: This is most likely to change after this https://github.com/WordPress/gutenberg/issues/30082#issuecomment-816679803

Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php]: I expect that `gutenberg_get_block_editor_settings` is going to be the primary way to get the settings. The question is how to pass the context - the name of the editor. It looks like settings are very different for the site editor, post editor or the widgets editor.

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-block-editor-settings-controller.php]: This needs to check for other any post type that might not use the `edit_posts` cap. For instance, from the themes controller.

```
		if ( current_user_can( 'edit_posts' ) ) {
			return true;
		}

		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
			if ( current_user_can( $post_type->cap->edit_posts ) ) {
				return true;
			}
		}
```

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-block-editor-settings-controller.php]:
```suggestion
			return new WP_Error( 'rest_cannot_read_block_editor_settings', $error, array( 'status' => rest_authorization_required_code() ) );
```

Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php]: Thanks for the suggestion, I updated it 👍

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php]: Awesome, thanks for that! Updated 🙇

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php]: Something like [this](https://github.com/WordPress/gutenberg/pull/29969/files#diff-593ccf28f1034efadc57f8992c879282300de88d6672c9ba4ab5f7764a307ad5R81)? Getting the `context` from a param and passing it to `gutenberg_get_block_editor_settings`?

One question by the way, after updating with `trunk` I'm getting:

```
 Undefined index: imageSizes in /lib/editor-settings.php on line 22
```

From this [changes](https://github.com/WordPress/gutenberg/pull/30955/files), so now I'm wondering if I need to pass something else or there's something I'm missing 😅

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (gziolo) [inline lib/class-wp-rest-block-editor-settings-controller.php]: `gutenberg_get_block_editor_settings` has `block_editor_setting` filter internally.

It should be fine to run `gutenberg_get_block_editor_settings` only.

Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php]: Thanks! That solves the issue I mentioned [here](https://github.com/WordPress/gutenberg/pull/29969#discussion_r616708627) 😅  

I still had it because I wanted to have the global styles settings but I'll wait until this [work](https://github.com/WordPress/gutenberg/issues/30082#issuecomment-816679803) lands. So I'll remove that 👍

Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php]: If a plugin defines a filter that uses `$post` object in the callback passed to `block_editor_settings` then it will throw an error. 

It's all covered in `gutenberg_get_block_editor_settings`:

https://github.com/WordPress/gutenberg/blob/a88a3b587cf203c62c196332d64b33afe164d283/lib/block-editor.php#L277-L290

**Review by TimothyBJacobs — APPROVED** [review]
Looks great to me, thanks for adding the Schema!

**Comment by gziolo** [comment]
@TimothyBJacobs - this endpoint will return a set of common properties, but when you pass an editor type like `post-editor` or `widgets-editor` to the API call then it will have also some properties specific only to the given type. I don't know the details at the moment, but it's something that we would have to take into account when building the schema.

**Comment by TimothyBJacobs** [comment]
In the schema we can register all the different possible settings. 

If we want, we can also use `context` to specify in which Gutenberg contexts that field is available. So instead of using `view` context would be like `[ post-editor, site-editor, widget-editor ]`, etc...

**Comment by gziolo** [comment]
> In the schema we can register all the different possible settings.
> 
> If we want, we can also use `context` to specify in which Gutenberg contexts that field is available. So instead of using `view` context would be like `[ post-editor, site-editor, widget-editor ]`, etc...

@TimothyBJacobs, thanks for the clarification. This approach would be great 👍🏻

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (gziolo) [inline lib/class-wp-rest-block-editor-settings-controller.php]: @youknowriad, see this context which is meant to allow the mobile app to pick settings for a given block editor type. As discussed in https://github.com/WordPress/gutenberg/pull/31027#issuecomment-824151592, we need to figure out what those contexts should be.

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (oandregal) [inline lib/class-wp-rest-block-editor-settings-controller.php:96]: This sounds great. It looks like there are some items that are off in terms of in which context they're available. Some that I've noticed that should be `site-editor` only are: `__experimentalGlobalStylesUserEntityId`, `__experimentalGlobalStylesBaseStyles`. Other than that, it seems that we can land this PR already? The global styles settings use the old filter `block_editor_settings` and we should update it but it doesn't block this PR, as far as I'm aware.

Also: any thoughts on best practices to keep this schema up to date as things evolve would be great.

Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php:96]: > Also: any thoughts on best practices to keep this schema up to date as things evolve would be great.

If we would start using the same endpoint for the web then it could help to maintain the schema 😄 

In related efforts, I plan we switch to the REST API endpoint for block types on the web rather than expose them with JS function call on the server.

Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php:96]: Thanks @nosolosw !

> Some that I've noticed that should be site-editor only are: __experimentalGlobalStylesUserEntityId, __experimentalGlobalStylesBaseStyles

I updated those to be `site-editor` only.

> The global styles settings use the old filter block_editor_settings and we should update it but it doesn't block this PR, as far as I'm aware.

Awesome 😃! Do you know more or less when that's going to be updated?

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php]: @draganescu, should it be `widget-editor` or `widgets-editor`?

**Comment by geriux** [comment]
Hey @nosolosw @gziolo 👋 I was wondering what are the next steps to get this merged. Thanks!

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php]: > @draganescu, should it be `widget-editor` or `widgets-editor`?

@talldan, @noisysocks, @tellthemachines, or @kevin940726 - can you help?

**Comment by gziolo** [comment]
There is one minor question I have https://github.com/WordPress/gutenberg/pull/29969#discussion_r620977651. Once it's confirmed we can land.

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (talldan) [reply lib/class-wp-rest-block-editor-settings-controller.php]: `widgets-editor` seems to be preferred everywhere. Does this match another value defined in the codebase? The name of the editor passed to `gutenberg_initialize_editor` is currently `widgets_editor` with an underscore.

There's also the customizer screen, which is `widgets_customizer`.

**Review by gziolo — APPROVED** [review]
I applied some changes to the REST API schema in https://github.com/WordPress/gutenberg/pull/29969/commits/3897d19f7e41a81f11b98e2337de81169dab8949 and I think everything looks good now. Let's iterate on it as we apply changes to the block editor setting while exposing new features planned for WordPress 5.8.

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php]: `widgets-editor` it is then. I updated the code in https://github.com/WordPress/gutenberg/pull/29969/commits/3897d19f7e41a81f11b98e2337de81169dab8949. Thank you.

> There's also the customizer screen, which is widgets_customizer.

Will it use a different set of settings?

**Comment by geriux** [comment]
> I applied some changes to the REST API schema in [3897d19](https://github.com/WordPress/gutenberg/commit/3897d19f7e41a81f11b98e2337de81169dab8949) and I think everything looks good now. Let's iterate on it as we apply changes to the block editor setting while exposing new features planned for WordPress 5.8.

Awesome! Thank you!

**Comment by youknowriad** [comment]
Does this mean we should revisit the way we initialize the editor? Instead of receiving settings from php, call an API to retrieve them?

**Comment by gziolo** [comment]
> Does this mean we should revisit the way we initialize the editor? Instead of receiving settings from php, call an API to retrieve them?

I think we should explore that approach for block editor settings and block types now that we have REST API endpoints that are capable of doing that.

**Comment by oandregal** [comment]
Created a follow-up for this at https://github.com/WordPress/gutenberg/pull/31534

It also looks like we need more structured info through the REST API for mobile (styles). Going to prepare a PR for that soon (hopefully today).

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (chrisvanpatten) [inline lib/class-wp-rest-block-editor-settings-controller.php:158]: I've noted this in other PRs and on Trac, but the lack of post type context constitutes a large functional break for all of the Gutenberg sites I've worked on, and I expect for many others as well. Being able to restrict certain blocks to certain post types is a crucial feature of many advanced block editor implementations.

Comment (gziolo) [reply lib/class-wp-rest-block-editor-settings-controller.php:158]: We discussed with @chrisvanpatten in https://github.com/WordPress/wordpress-develop/pull/1118#discussion_r627396445 that it should be considered passing additional params in addition to the context to bring feature parity with how the edit post editor works today. It looks like the possibility to change settings per individual post/page is used by the community. It means that we should explore how params like `post_id` could be passed and consumed by this endpoint.

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/31762 for passing consolidated structured styles for the mobile apps

File: lib/class-wp-rest-block-editor-settings-controller.php
Comment (oandregal) [reply lib/class-wp-rest-block-editor-settings-controller.php:96]: I've just merged https://github.com/WordPress/gutenberg/pull/31762 which adds more data for mobile. Upon hitting the button, it just occurred to me that something was off. I think it boils down to the contexts we define in this PR: `post-editor`, `site-editor`, `widgets-editor`. Shouldn't we also have a `mobile` context?

My understanding is that we use this to filter what kind of data we pass through the request. If the plan is to move the settings retrieval to use the REST request, checks like https://github.com/WordPress/gutenberg/pull/31762/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR91 don't make sense. However, the current contexts are also not enough. I'm thinking we actually need a `mobile` context as well. @geriux et all, do you have thoughts on this?

Comment (geriux) [reply lib/class-wp-rest-block-editor-settings-controller.php:96]: > I'm thinking we actually need a mobile context as well. @geriux et all, do you have thoughts on this?

I think we might if we want specific data like the consolidated styles we added [here](https://github.com/WordPress/gutenberg/pull/31762/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR106) since the web editor doesn't look like it'd need that right?

Comment (oandregal) [reply lib/class-wp-rest-block-editor-settings-controller.php:96]: OK, prepared https://github.com/WordPress/gutenberg/pull/31819 to add a `mobile` context.

**Pull request merged**
